service/src/ConstConfigIOGroup.cpp

```diff
@@ -0,0 +1,424 @@
+    {
+        return GEOPM_DOMAIN_INVALID;
+    }
+
+    int ConstConfigIOGroup::push_signal(const std::string &signal_name,
+                                        int domain_type, int domain_idx)
+    {
+        auto it = m_signal_available.find(signal_name);
+        if (it == m_signal_available.end()) {
+            throw Exception("ConstConfigIOGroup::push_signal(): " +
+                            signal_name + " not valid for ConstConfigIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        else if (domain_type != it->second.domain) {
+            throw Exception("ConstConfigIOGroup::push_signal(): domain_type " +
+                            std::to_string(domain_type) +
+                            "not valid for ConstConfigIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        else if (domain_idx < 0 || domain_idx >= it->second.values.size()) {
+            throw Exception("ConstConfigIOGroup::push_signal(): domain_idx " +
+                            std::to_string(domain_idx) + "out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        size_t signal_idx = std::distance(m_signal_available.begin(), it);
+        bool is_found = false;
+        for (auto const &signal : m_pushed_signals) {
+            if (signal_idx == signal.signal_idx &&
+                domain_idx == signal.domain_idx) {
+                is_found = true;
+                break;
+            }
+        }
+
+        if (!is_found) {
+            m_signal_ref_s signal = {
+                .signal_idx = signal_idx,
+                .domain_idx = domain_idx
+            };
+            m_pushed_signals.push_back(signal);
+        }
+        
+        return m_pushed_signals.size() - 1;
+    }
+
+    int ConstConfigIOGroup::push_control(const std::string &control_name,
+                                         int domain_type, int domain_idx)
+    {
+        throw Exception("ConstConfigIOGroup::push_control(): there are no "
+                        "controls supported by the ConstConfigIOGroup",
+                        GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+    }
+
+    void ConstConfigIOGroup::read_batch(void)
+    {
+    }
+
+    void ConstConfigIOGroup::write_batch(void)
+    {
+    }
+
+    double ConstConfigIOGroup::sample(int batch_idx)
+    {
+        double result = NAN;
+        
+        if (batch_idx < 0 || batch_idx >= m_pushed_signals.size()) {
+            throw Exception("ConstConfigIOGroup::sample(): batch_idx " +
+                            std::to_string(batch_idx) + " out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        m_signal_ref_s signal = m_pushed_signals[batch_idx];
+        auto it = m_signal_available.begin();
+        std::advance(it, signal.signal_idx);
+        result = it->second.values[signal.domain_idx];
```

Comment:
I think this only works if available signals never changes in between calls to this (which would potentially change the offsets as new keys are inserted/removed). Looks your current implementation does avoid that.

Should we enforce that by making `m_signal_available` a `const` member? You'd need to initialize it in the constructor instead of post-construction. E.g., `ConstConfigIOGroup(const std::string &config) : m_signal_available(something_that_returns_the_map(config))`

Reply:
I'm in favor of this suggestion.

Reply:
Yeah, I'm not fully convinced with my initial implementation. I'll revisit this soon. If I do stick with this implementation, I'll make sure to follow your suggestion of making ``m_signal_available`` a const member.